agent/csi/libstorage/csi_service_libstorage.go

```diff
@@ -381,15 +445,94 @@ func (d *driver) NodePublishVolume(
 			"invalid volume capability request"), nil
 	}
 
-	return nil, nil
+	if mv := req.VolumeCapability.GetMount(); mv != nil {
+		return d.handleMountVolume(
+			volumeID, target, mv, mv.MountFlags, vi, req.Readonly)
+	}
+	if bv := req.VolumeCapability.GetBlock(); bv != nil {
+		return d.handleBlockVolume(target, vi)
+	}
+
+	return gocsi.ErrNodePublishVolume(
+		csi.Error_NodePublishVolumeError_UNSUPPORTED_VOLUME_TYPE,
+		"No supported volume type received"), nil
 }
 
 func (d *driver) NodeUnpublishVolume(
 	ctx xctx.Context,
 	req *csi.NodeUnpublishVolumeRequest) (
 	*csi.NodeUnpublishVolumeResponse, error) {
 
-	return nil, nil
+	volumeID, ok := req.VolumeId.Values["id"]
+	if !ok {
+		return gocsi.ErrNodeUnpublishVolume(
+			csi.Error_NodeUnpublishVolumeError_INVALID_VOLUME_ID,
+			`missing "id" field`), nil
+	}
+
+	target := req.GetTargetPath()
+
+	opts := &apitypes.VolumeInspectOpts{Opts: apiutils.NewStore()}
+	opts.Attachments = apitypes.VolAttReqWithDevMapOnlyVolsAttachedToInstance
+	vol, err := d.client.Storage().VolumeInspect(d.ctx, volumeID, opts)
+	if err != nil {
+		return nil, err
+	}
+	if vol == nil {
```

Comment:
Although this is something that needs to be fixed, as I just tried it out (hadn't done that) where I call a `NodeUnpublish` *after* a succesful `ControllerUnpublish`, which is something that shouldn't happen per the spec. But if you do, this logic fails because you get an HTTP 404 not found from libStorage, and we need to check for that -- it is an expected error.